subsys/mgmt/fmfu/src/mgmt_fmfu.c

```diff
@@ -0,0 +1,281 @@
+	err |= cbor_encode_uint(&ctx->encoder, expected_offset);
+
+	if (err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+}
+
+static int mgmt_firmware_upload(struct mgmt_ctxt *ctx)
+{
+
+	int ret;
+	int err;
+	bool whole_file_received;
+	static struct firmware_packet packet;
+	uint32_t next_expected_offset;
+
+	int modem_state = nrf_fmfu_modem_state_get();
+	static bool first_chunk = true;
+
+	LOG_INF("Modem state: %d", modem_state);
+	ret = unpack(ctx, &packet, &whole_file_received);
+	if (ret < 0) {
+		return ret;
+	}
+
+	next_expected_offset = ret;
+	LOG_INF("New offset: 0x%x", next_expected_offset);
+
+	if (packet.data_len > 0) {
+		if (first_chunk) {
+			err = nrf_fmfu_transfer_start();
+			if (err != 0) {
+				return get_mgmt_err_from_modem_ret_err(err);
+			}
+			first_chunk = false;
+		}
+
+		LOG_INF("target_addr: 0x%x", packet.target_address);
+		LOG_INF("Writing firmware packet with len:%d", packet.data_len);
+		err = nrf_fmfu_memory_chunk_write(packet.target_address,
+						  packet.data_len,
+						  &packet.data);
+		if (err != 0) {
+			LOG_INF("Error in writing data");
+			return get_mgmt_err_from_modem_ret_err(err);
+		}
+	}
+
+	if (whole_file_received) {
+		LOG_INF("whole file recived");
+		err = nrf_fmfu_transfer_end();
+		if (err != 0) {
+			LOG_INF("error in transfer_end");
+			return get_mgmt_err_from_modem_ret_err(err);
+		}
+		first_chunk = true;
+	}
+
+	return encode_response(ctx, next_expected_offset);
+}
+
+static int mgmt_get_memory_hash(struct mgmt_ctxt *ctxt)
+{
+	unsigned long long start;
+	unsigned long long end;
+	struct nrf_fmfu_digest digest;
+
+	/* We expect two variables: the start and end address of a memory region
+	 * that we want to verify (obtain a hash for)
+	 */
+	const struct cbor_attr_t off_attr[] = {
+		[0] = {
+			.attribute = "start",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &start,
+			.nodefault = true
+		},
+		[1] = {
+			.attribute = "end",
+			.type = CborAttrUnsignedIntegerType,
+			.addr.uinteger = &end,
+			.nodefault = true
+		},
+		[2] = { 0 },
+	};
+	start = 0;
+	end = 0;
+
+	/* Parse JSON object */
+	int rc = cbor_read_object(&ctxt->it, off_attr);
+
+	if (rc || start == 0 || end == 0) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	int err = nrf_fmfu_memory_hash_get(start, end, &digest);
+
+	if (err != 0) {
+		return get_mgmt_err_from_modem_ret_err(err);
+	}
+
+	/* Put the digest response in the response */
+	CborError cbor_err = 0;
+
+	cbor_err |= cbor_encode_text_stringz(&ctxt->encoder, "digest");
+	cbor_err |= cbor_encode_byte_string(&ctxt->encoder,
+			(uint8_t *)digest.data, NRF_FMFU_DIGEST_BUFFER_LEN);
+	cbor_err |= cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	cbor_err |= cbor_encode_int(&ctxt->encoder, err);
+	if (cbor_err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return MGMT_ERR_EOK;
+}
+
+#define FIRMWARE_SHA_LENGTH 32
+static int unpack(struct mgmt_ctxt *ctxt, struct firmware_packet *packet,
+		  bool *whole_file_received)
+{
+	unsigned long long offset;
```

Comment:
Does these need to be `long long` ? I know they were this way, we should just double check why there are not simply ints.

Reply:
compiler warning. basically. I can type cast it the other way around.